### PR TITLE
Implemented Certification Declaration Generation/Verification APIs

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
SDK currently doesn't support Certification Declaration processing.

Note that these APIs are not used by devices but might be used by commissioner. 

#### Change overview
 -- Implemented APIs to do TLV encoding/decoding of certification elements
 -- Implemented APIs to do CMS signing/verification of certification declaration payload
 -- Added unit tests

#### Testing
Implemented unit tests for the new APIs

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Comment by emargolis** [comment]
@msandstedt @mspang @andy31415 @Damian-Nordic @LuDuda PTAL

**Comment by woody-apple** [comment]
@andy31415 ?

**Comment by emargolis** [comment]
@msandstedt @mspang @andy31415 @Damian-Nordic @LuDuda @tcarmelveilleux  PTAL

**Comment by woody-apple** [comment]
@emargolis Sorry, conflicts now :(

File: src/credentials/CertificationDeclaration.cpp [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+    ReturnErrorOnFailure(reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_VendorId)));
+    ReturnErrorOnFailure(reader.Get(element));
+    VerifyOrReturnError(CanCastTo<uint16_t>(element), CHIP_ERROR_INVALID_TLV_ELEMENT);
+    certElements.VendorId = static_cast<uint16_t>(element);
+
+    ReturnErrorOnFailure(reader.Next(kTLVType_Array, ContextTag(kTag_ProductIds)));
+    ReturnErrorOnFailure(reader.EnterContainer(outerContainer2));
+
+    certElements.ProductIdsCount = 0;
+    while ((err = reader.Next(kTLVType_UnsignedInteger, AnonymousTag)) == CHIP_NO_ERROR)
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:120]: updated

File: src/credentials/CertificationDeclaration.cpp [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+namespace {
+
+CHIP_ERROR EncodeEncapsulatedContent(const ByteSpan & cdContent, ASN1Writer & writer)
+{
+    /**
+     * EncapsulatedContentInfo ::= SEQUENCE {
+     *   eContentType OBJECT IDENTIFIER pkcs7-data (1.2.840.113549.1.7.1),
+     *   eContent [0] EXPLICIT OCTET STRING cd_content }
+     */
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:177]: `err` and `exit` are used by `ASN1_` macros.

File: src/credentials/CertificationDeclaration.cpp [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+}
+
+CHIP_ERROR DecodeEncapsulatedContent(ASN1Reader & reader, ByteSpan & cdContent)
+{
+    /**
+     * EncapsulatedContentInfo ::= SEQUENCE {
+     *   eContentType OBJECT IDENTIFIER pkcs7-data (1.2.840.113549.1.7.1),
+     *   eContent [0] EXPLICIT OCTET STRING cd_content }
+     */
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:205]: `err` and `exit` are used by `ASN1_` macros.

File: src/credentials/CertificationDeclaration.cpp [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+{
+    /**
+     * SignerInfo ::= SEQUENCE {
+     *   version INTEGER ( v3(3) ),
+     *   subjectKeyIdentifier OCTET STRING,
+     *   digestAlgorithm OBJECT IDENTIFIER sha256 (2.16.840.1.101.3.4.2.1),
+     *   signatureAlgorithm OBJECT IDENTIFIER ecdsa-with-SHA256 (1.2.840.10045.4.3.2),
+     *   signature OCTET STRING }
+     */
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:239]: `err` and `exit` are used by `ASN1_` macros.

File: src/credentials/CertificationDeclaration.cpp [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+                    MutableByteSpan & signedMessage)
+{
+    /**
+     * CertificationDeclaration ::= SEQUENCE {
+     *   version INTEGER ( v3(3) ),
+     *   digestAlgorithm OBJECT IDENTIFIER sha256 (2.16.840.1.101.3.4.2.1),
+     *   encapContentInfo EncapsulatedContentInfo,
+     *   signerInfo SignerInfo }
+     */
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:354]: `err` and `exit` are used by `ASN1_` macros.

File: src/credentials/CertificationDeclaration.cpp [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+    }
+    ASN1_EXIT_SEQUENCE;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CMS_ExtractKeyId(const ByteSpan & signedMessage, ByteSpan & signerKeyId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:481]: `err` and `exit` are used by `ASN1_` macros.

File: src/credentials/CertificationDeclaration.cpp [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+    }
+    ASN1_EXIT_SEQUENCE;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CMS_ExtractCDContent(const ByteSpan & signedMessage, ByteSpan & cdContent)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/credentials/CertificationDeclaration.cpp:524]: `err` and `exit` are used by `ASN1_` macros.

**Review by yufengwangca — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**